### DISCUSSION TITLE
Question: how to know what metric is used when doing a Buffer?

### DISCUSSION THREAD

**Discussion by PureKrome** [NetTopologySuite/NetTopologySuite#492] [Q&A]
Hello NTS,

I'm trying to create a bounding box around a specific POINT. To do this, I'll basically do a `Buffer(x).Envelope`.

I'm not sure sure what the _unit of measure_ `x` should be. So for example, If i want to make a 10km buffer, then I would assume i need to do a 5km `Buffer`.  Reading the info, it looks like the value is directly related to the `PrecisionModel` of the factory?

Given the 'Getting Started' docs...

```
NetTopologySuite.NtsGeometryServices.Instance = new NetTopologySuite.NtsGeometryServices(
    NetTopologySuite.Geometries.Implementation.CoordinateArraySequenceFactory.Instance,
    new NetTopologySuite.Geometries.PrecisionModel(1000d),
    4326,
    NetTopologySuite.Geometries.GeometryOverlay.NG);
```

the `PrecisionModel` is `1000` <something> for the SRID 4326 (GPS, etc)

Is this 1000 meters (1 km) so if I start doing `Buffer(5)` that will now be 5km?

**Comment by FObermaier** [comment]
Same as coordinates, in case of SRID=4326 degrees, unfortunately.
In NTS all is planar.

**Comment by PureKrome** [comment]
So is there an _easy_ way to say `Buffer(5km)` then?

**Comment by DGuidi** [comment]
as felix explained, the unit of measure depends of the SRID you set; so if you use as example an UTM SRID like [32632](https://epsg.io/32632) you are working with *metres*, so if you want a buffer of 5 km around a point you can use `Buffer(5000)`.    
The key part to understand is that the units in NTS are **just numbers**:  if you use a *global* srid like [4326](https://epsg.io/4326) you are working with **degrees**; NTS doesn't account for the earth curvature or other kinds of corrections, and you can not expect to use `Buffer(5000)` to have a buffer of 5 km around your point.   
Take a look at #443 too, maybe can be helpful.

**Comment by PureKrome** [comment]
Cheers - that helps .. and makes me cry 😢 

Also, I was just assuming that the `operations on the `Geometry` class actually took the curvature of the earth into account. Sure, i'm dealing with military grade calculations and precision but I just assumed that this was the default. urgh.

It hurts so much how this is all so complicated.

So -> if I wish to do a 5km circle around a lat/long, i'm now going to have to do something like this?

- create a point with WSG84 `4326`
- convert this point to URM 32632
- `Buffer(5000)`
- convert back to WSG84 ?

I've got some older code (before I was looking at using NTS) which was ripping some code from somewhere which is the Haversine Formula .. which sorta looks ok ... but of course, was just seeing if there was a built in way with NTS.

**Comment by airbreather** [comment]
See also #233

> So -> if I wish to do a 5km circle around a lat/long, i'm now going to have to do something like this?
> * create a point with WSG84 `4326`
> * convert this point to URM 32632
> * `Buffer(5000)`
> * convert back to WSG84 ?

Pretty much, yeah, though the exact ideal projection to use depends on your data.

You also ought to consider stupid cases like antimeridian crossings and projection bounds, unless you can hand-wave the issues away by restricting the inputs.  I rant a little bit about this kind of thing in https://github.com/NetTopologySuite/NetTopologySuite/issues/233#issuecomment-455780259, and that's not even the end of it: if your buffer happens to cover either the North or South Pole, then most algorithms in NTS will probably break down.

**Comment by PureKrome** [comment]
Oh wow - love that rant! Personally, I don't see it as a rant because I'm feeling lots of what you said. Sure, I don't deal with polys that cross the antimeridian and even between hemisphere's (hi sql server 2008 "denali" when geography first came out) ... but I do deal with (what I would have thought was) simple, common spatial operations:

- lat/longs
- buffers (in meters or kms)
- point in poly (intersect)
- bounding box (envelope) 
- distance between two points
- centre point of a poly (which is so much visual fun, when the poly is a weird shape ... and even more fun when the centre point lies outside of the weird ass shape!)

I also hated the fact that with SqlServer they have 2 types of types: `Geom` and `Geog` ... like .. they feel like the same thing most of the time .. until you want to do an operation that is either 3D (world sphere) or 2D. ... so I just ignorantly thought that the NTS `Geometry` was a standard way and when you wanted to do some operations against a world sphere, you could just set a param so the underlying magic might just 'know'.

I'm not a map person at all. Totally newbie. But i just would have thought that if _anyone_ was going to do _anything_ with a Lat/Long .. that basically says they are doing something with the Earth. So therefore, any WSG84 4326 operations would take into consideration the curvature of the Earth. So my lack of education here (and gross assumption) probably insults very cartographer out there <blush> .

But back to the topic -> I guess i can experiment with the above idea of projecting back and forth.

And again, i thought this would be like .. _really common stuff_ for people ... that this could (should?) be documented somewhere, without getting into a deep dive about projections and stuff?

**Comment by jkuek** [comment]
I've run into the same problem just recently. I want to look at the location reported by a device and determine when it comes in/out of a defined geofence (which is a point + radius). I would imagine this is a very common use-case.

My device location and geofence coordinates are both in WGS84, but the geofence radius (if it is a circle) is in meters. I also have an accuracy measure from the device which is in meters.

The accepted way using NTS seems to be to:

1. transform both location and geofence to a suitable projection that has units of meters - I tried WGS84_UTM
2. Add .Buffer(radius) to transformed geofence point
3. calculate the distance between the transformed location and transformed geofence e.g. geofence.Distance(location)
4. compare this calculated distance with location accuracy

While this did give me the results I wanted, it felt a bit clunky (perhaps because I'm not so used to dabbling in this space and NTS is new to me)

I opted for a slightly different approach:

1. Use a DistanceOp to find the closest points between location and geofence
2. Use Haversine formula to calculate the distance between those points
3. Compare that distance to (location accuracy + geofence radius) to see if location is in or out

Are there any gotchas with this second approach? To me, the code looks much neater 

Example code snippet

```
double maxDistance = location.Accuracy.GetValueOrDefault();

geofenceGeometry = new Point(new Coordinate(geofence.Longitude, geofence.Latitude));

maxDistance += geofence.Radius;

Point locationPoint = new Point(location.Longitude, location.Latitude);

DistanceOp distOp = new DistanceOp(locationPoint, geofenceGeometry);
var nearestPoints = distOp.NearestPoints();

var distance = HaversineDistance(nearestPoints[0], nearestPoints[1]);

return (distance < maxDistance);
```
